net/tcl-dp/man/man3/Address.3
'\"
' $Header: /a/cvs/386BSD/ports/net/tcl-dp/man/man3/Address.3,v 1.1 1993/09/01 00:46:31 jkh Exp $ SPRITE (Berkeley)
.\" The definitions below are for supplemental macros used in Sprite
.\" manual entries.
.\"
.\" .HS name section [date [version]]
.\"	Replacement for .TH in other man pages.  See below for valid
.\"	section names.
.\"
.\" .AP type name in/out [indent]
.\"	Start paragraph describing an argument to a library procedure.
.\"	type is type of argument (int, etc.), in/out is either "in", "out",
.\"	or "in/out" to describe whether procedure reads or modifies arg,
.\"	and indent is equivalent to second arg of .IP (shouldn't ever be
.\"	needed;  use .AS below instead)
.\"
.\" .AS [type [name]]
.\"	Give maximum sizes of arguments for setting tab stops.  Type and
.\"	name are examples of largest possible arguments that will be passed
.\"	to .AP later.  If args are omitted, default tab stops are used.
.\"
.\" .BS
.\"	Start box enclosure.  From here until next .BE, everything will be
.\"	enclosed in one large box.
.\"
.\" .BE
.\"	End of box enclosure.
.\"
.\" .VS
.\"	Begin vertical sidebar, for use in marking newly-changed parts
.\"	of man pages.
.\"
.\" .VE
.\"	End of vertical sidebar.
.\"
.\" .DS
.\"	Begin an indented unfilled display.
.\"
.\" .DE
.\"	End of indented unfilled display.
.\"
'	# Heading for Sprite man pages
.de HS
.if '\\$2'cmds'       .TH \\$1 1 \\$3 \\$4
.if '\\$2'lib'        .TH \\$1 3 \\$3 \\$4
.if '\\$2'tcl'        .TH \\$1 3 \\$3 \\$4
.if '\\$2'tk'         .TH \\$1 3 \\$3 \\$4
.if t .wh -1.3i ^B
.nr ^l \\n(.l
.ad b
..
'	# Start an argument description
.de AP
.ie !"\\$4"" .TP \\$4
.el \{\
.   ie !"\\$2"" .TP \\n()Cu
.   el          .TP 15
.\}
.ie !"\\$3"" \{\
.ta \\n()Au \\n()Bu
\&\\$1	\\fI\\$2\\fP	(\\$3)
.\".b
.\}
.el \{\
.br
.ie !"\\$2"" \{\
\&\\$1	\\fI\\$2\\fP
.\}
.el \{\
\&\\fI\\$1\\fP
.\}
.\}
..
'	# define tabbing values for .AP
.de AS
.nr )A 10n
.if !"\\$1"" .nr )A \\w'\\$1'u+3n
.nr )B \\n()Au+15n
.\"
.if !"\\$2"" .nr )B \\w'\\$2'u+\\n()Au+3n
.nr )C \\n()Bu+\\w'(in/out)'u+2n
..
'	# BS - start boxed text
'	# ^y = starting y location
'	# ^b = 1
.de BS
.br
.mk ^y
.nr ^b 1u
.if n .nf
.if n .ti 0
.if n \l'\\n(.lu\(ul'
.if n .fi
..
'	# BE - end boxed text (draw box now)
.de BE
.nf
.ti 0
.mk ^t
.ie n \l'\\n(^lu\(ul'
.el \{\
.\"	Draw four-sided box normally, but don't draw top of
.\"	box if the box started on an earlier page.
.ie !\\n(^b-1 \{\
\h'-1.5n'\L'|\\n(^yu-1v'\l'\\n(^lu+3n\(ul'\L'\\n(^tu+1v-\\n(^yu'\l'|0u-1.5n\(ul'
.\}
.el \}\
\h'-1.5n'\L'|\\n(^yu-1v'\h'\\n(^lu+3n'\L'\\n(^tu+1v-\\n(^yu'\l'|0u-1.5n\(ul'
.\}
.\}
.fi
.br
.nr ^b 0
..
'	# VS - start vertical sidebar
'	# ^Y = starting y location
'	# ^v = 1 (for troff;  for nroff this doesn't matter)
.de VS
.mk ^Y
.ie n 'mc \s12\(br\s0
.el .nr ^v 1u
..
'	# VE - end of vertical sidebar
.de VE
.ie n 'mc
.el \{\
.ev 2
.nf
.ti 0
.mk ^t
\h'|\\n(^lu+3n'\L'|\\n(^Yu-1v\(bv'\v'\\n(^tu+1v-\\n(^Yu'\h'-|\\n(^lu+3n'
.sp -1
.fi
.ev
.\}
.nr ^v 0
..
'	# Special macro to handle page bottom:  finish off current
'	# box/sidebar if in box/sidebar mode, then invoked standard
'	# page bottom macro.
.de ^B
.ev 2
'ti 0
'nf
.mk ^t
.if \\n(^b \{\
.\"	Draw three-sided box if this is the box's first page,
.\"	draw two sides but no top otherwise.
.ie !\\n(^b-1 \h'-1.5n'\L'|\\n(^yu-1v'\l'\\n(^lu+3n\(ul'\L'\\n(^tu+1v-\\n(^yu'\h'|0u'\c
.el \h'-1.5n'\L'|\\n(^yu-1v'\h'\\n(^lu+3n'\L'\\n(^tu+1v-\\n(^yu'\h'|0u'\c
.\}
.if \\n(^v \{\
.nr ^x \\n(^tu+1v-\\n(^Yu
\kx\h'-\\nxu'\h'|\\n(^lu+3n'\ky\L'-\\n(^xu'\v'\\n(^xu'\h'|0u'\c
.\}
.bp
'fi
.ev
.if \\n(^b \{\
.mk ^y
.nr ^b 2
.\}
.if \\n(^v \{\
.mk ^Y
.\}
..
'	# DS - begin display
.de DS
.RS
.nf
.sp
..
'	# DE - end display
.de DE
.fi
.RE
.sp .5
..
.HS Address tk
.BS
.SH NAME
Address \- C level access to Tcl-DP address records
.SH SYNOPSIS
.nf
.sp
\fB#include "network.h"\fR 
.sp
struct sockaddr_in *
\fBTcm_FindAddress (\fIname\fB)\fR
.sp
char *
\fBTcm_CreateAddress (\fIaddr, port\fB)\fR
.SH ARGUMENTS
.AS unsigned_long name
.AP char *name in
An address handle.
.AP unsigned_long addr in
An inet address.
.AP int port in
A port number.
.BE
.SH DESCRIPTION
.PP

The above procedures provide C access to the Tcl-DP "address"
extension.  The procedure \fBTcm_CreateAddress\fR creates a
record to store the given \fIaddr\fR and \fIport\fR information
and returns an address handle for further access to the created
record.  The \fBTcm_FindAddress\fR procedure returns an address
record given an address handle.
.SH AUTHOR
Brian Smith (bsmith@cs.berkeley.edu)
.br
Steve Yen (syen@cs.berkeley.edu)


net/tcl-dp/address.c
/*
 * address.c
 *
 *	This file implements the "address" extension of Tcl-DP.
 *
 * Copyright 1992 Regents of the University of California.
 * Permission to use, copy, modify, and distribute this
 * software and its documentation for any purpose and without
 * fee is hereby granted, provided that the above copyright
 * notice appear in all copies.  The University of California
 * makes no representations about the suitability of this
 * software for any purpose.  It is provided "as is" without
 * express or implied warranty.
 */

#include <stdio.h>
#include <sys/types.h>
#include <sys/time.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include "tcl.h"

/*
 * One record of the following type is kept for each inetAddr:
 */

typedef struct InetAddr {
    char name[10];
    struct sockaddr_in addr;
    struct InetAddr *next;
} InetAddr;

static InetAddr *addrList;

/*
 *--------------------------------------------------------------
 *
 * Tcm_CreateAddress --
 *
 *	Create and initialize an inet address struct.  This routine can be
 *	called by C routines and guarantees that if the inet address 
 *	struct already exists for a given addr and port, a new 
 *	struct won't be created.
 *
 * Results:
 *	Handle (an address handle) to created inet address struct.
 *
 * Side effects:
 *	Memory is allocated.
 *
 *--------------------------------------------------------------
 */
char *
Tcm_CreateAddress(addr, port)
    unsigned long addr;		/* Address as a 4 byte int. */
    int port;			/* Port number (host order) */
{
    static int addrNum;
    InetAddr *newAddr;
    InetAddr *searchAddr;

    /*
     * First, see if it already exists
     */
    for (searchAddr = addrList; searchAddr != NULL; searchAddr = searchAddr->next) {
	if ((searchAddr->addr.sin_addr.s_addr == addr) &&
	    (searchAddr->addr.sin_port == htons(port))) {
	    return (searchAddr->name);
	}
    }

    /*
     * Alloc it, give it a name, and link it into the list.
     */
    newAddr = (InetAddr *) ckalloc(sizeof(InetAddr));
    sprintf(newAddr->name, "addr%d", addrNum++);
    newAddr->addr.sin_family = AF_INET;
    newAddr->addr.sin_addr.s_addr = addr;
    newAddr->addr.sin_port = htons(port);
    newAddr->next = addrList;
    addrList = newAddr;

    return (newAddr->name);
}

/*
 *--------------------------------------------------------------
 *
 * Tcm_FindAddr --
 *
 *	Find and return the internet address structure with the
 *	associated name.
 *
 * Results:
 *	Pointer to the internet structure for the address, or NULL
 *	if it can't be found.
 *
 * Side effects:
 *	None.
 *
 *--------------------------------------------------------------
 */
struct sockaddr *
Tcm_FindAddr(name)
    char *name;
{
    InetAddr *addrPtr;

    /*
     * Find the entry (error if there isn't one).
     */
    for (addrPtr = addrList; addrPtr != NULL; addrPtr = addrPtr->next) {
	if (strcmp(addrPtr->name, name) == 0) {
	    return (struct sockaddr *)(&addrPtr->addr);
	}
    }
    return ((struct sockaddr *) NULL);
}

/*
 *--------------------------------------------------------------
 *
 * Tcm_AddressCmd --
 *
 *	This procedure is invoked to process the "address" Tcl/Tk
 *	command.  See the user documentation for details on what
 *	it does.
 *
 * Results:
 *	A standard Tcl result.
 *
 * Side effects:
 *	See the user documentation.
 *
 *--------------------------------------------------------------
 */
 /* ARGSUSED */
int
Tcm_AddressCmd(notUsed, interp, argc, argv)
    ClientData notUsed;
    Tcl_Interp *interp;		/* Current interpreter. */
    int argc;			/* Number of arguments. */
    char **argv;		/* Argument strings. */
{
    char c;
    int len;

    if (argc < 2) {
	Tcl_AppendResult(interp, "wrong # args: should be \"",
			 argv[0], " option ?arg arg ...?\"", (char *) NULL);
	return TCL_ERROR;
    }
    c = argv[1][0];
    len = strlen(argv[1]);

    /*------------------------ CREATE -----------------------------*/
    if ((c == 'c') && (strncmp(argv[1], "create", len) == 0)) {
	struct hostent *hp;
	unsigned long addr;
	int port;
	char *name;

	/*
	 * Check syntax and validity of args.
	 */
	if (argc != 4) {
	    Tcl_AppendResult(interp, "wrong # args: should be \"",
			     argv[0], " create host port\"", (char *) NULL);
	    return TCL_ERROR;
	}
	hp = gethostbyname(argv[2]);
	if (hp == NULL) {
	    addr = inet_addr(argv[2]);
	    if (addr == -1) {
		Tcl_AppendResult(interp, argv[0], " create: unknown host \"",
				 argv[2], "\"", (char *) NULL);
		return TCL_ERROR;
	    }
	} else {
	    bcopy((char *) hp->h_addr_list[0], (char *) &addr, sizeof(addr));
	}
	if (Tcl_GetInt(interp, argv[3], &port) != TCL_OK) {
	    return TCL_ERROR;
	}
	/*
	 * Looks good -- alloc a new struct and initialize it.
	 */
	name = Tcm_CreateAddress(addr, port);
	Tcl_AppendResult(interp, name, (char *) NULL);

    /*------------------------ DELETE -----------------------------*/
    } else if ((c == 'd') && (strncmp(argv[1], "delete", len) == 0)) {
	InetAddr *prevPtr;
	InetAddr *addrPtr;

	/*
	 * Check args
	 */
	if (argc != 3) {
	    Tcl_AppendResult(interp, "wrong # args: should be \"",
			     argv[0], " delete address\"", (char *) NULL);
	    return TCL_ERROR;
	}
	/*
	 * Find the entry (error if there isn't one).
	 */
	for (prevPtr = NULL, addrPtr = addrList;;
	     prevPtr = addrPtr, addrPtr = addrPtr->next) {
	    if (addrPtr == NULL) {
		Tcl_AppendResult(interp, argv[0], " delete: invalid address \"",
				 argv[2], "\"", (char *) NULL);
		return TCL_ERROR;
	    }
	    if (strcmp(addrPtr->name, argv[2]) == 0) {
		break;
	    }
	}

	/*
	 * Delete this one from the list.
	 */
	if (prevPtr == NULL) {
	    addrList = addrPtr->next;
	} else {
	    prevPtr->next = addrPtr->next;
	}
	free((char *) addrPtr);

    /*------------------------ INFO -------------------------------*/
    } else if ((c == 'i') && (strncmp(argv[1], "info", len) == 0)) {
	InetAddr *addrPtr;
	int f1, f2, f3, f4;

	/*
	 * Check args
	 */
	if (argc != 3) {
	    Tcl_AppendResult(interp, "wrong # args: should be \"",
			     argv[0], " info address\"", (char *) NULL);
	    return TCL_ERROR;
	}
	/*
	 * Find the entry (error if there isn't one).
	 */
	for (addrPtr = addrList;; addrPtr = addrPtr->next) {
	    if (addrPtr == NULL) {
		Tcl_AppendResult(interp, argv[0], " info: invalid address \"",
				 argv[2], "\"", (char *) NULL);
		return TCL_ERROR;
	    }
	    if (strcmp(addrPtr->name, argv[2]) == 0) {
		break;
	    }
	}

	/*
	 * Get the info from this one...
	 */
#ifdef BIG_ENDIAN
	f1 = (addrPtr->addr.sin_addr.s_addr >> 24) & 0xff;
	f2 = (addrPtr->addr.sin_addr.s_addr >> 16) & 0xff;
	f3 = (addrPtr->addr.sin_addr.s_addr >> 8) & 0xff;
	f4 = addrPtr->addr.sin_addr.s_addr & 0xff;
#else
	f4 = (addrPtr->addr.sin_addr.s_addr >> 24) & 0xff;
	f3 = (addrPtr->addr.sin_addr.s_addr >> 16) & 0xff;
	f2 = (addrPtr->addr.sin_addr.s_addr >> 8) & 0xff;
	f1 = addrPtr->addr.sin_addr.s_addr & 0xff;
#endif
	sprintf(interp->result, "%d.%d.%d.%d %d", f1, f2, f3, f4,
		ntohs(addrPtr->addr.sin_port));
    } else {
	Tcl_AppendResult(interp, argv[0], "unknown option \"",
			 argv[1], "\"", (char *) NULL);
	return TCL_ERROR;
    }
    return TCL_OK;
}
